niceman/resource/singularity.py

```diff
@@ -0,0 +1,240 @@
+            # Running singularity instances and managing them didn't happen
+            # until version 2.4. See: https://singularity.lbl.gov/archive/
+            raise CommandError(msg="Singularity version >= 2.4 required.")
+
+        # Get instance info if we have one running.
+        info = self.get_instance_info()
+        if info:
+            if not self.name:
+                self.name = info['name']
+            if not self.id:
+                # Daemon names for a singularity instance are unique so they
+                # can do double duty as the ID. While not a global ID, it
+                # will do the trick locally.
+                self.id = info['name']
+            self.image = info['image']
+            self.status = 'running'
+        else:
+            self.id = None
+            self.status = None
+
+    def create(self):
+        """
+        Create a container instance.
+
+        Returns
+        -------
+        dict : config parameters to capture in the inventory file
+        """
+        # Check to see if the container is already running.
+        if self.get_instance_info():
+            lgr.info('Resource {} already exists.'.format(self.name))
+        else:
+            self.image_file = "{}.simg".format(str(uuid.uuid4().hex)[:16])
+            if self.image.startswith('shub://') or \
+                    self.image.startswith('docker://'):
+
+                self._runner.run(['singularity', 'pull', '--name',
+                    self.image_file, self.image])
+            else:
+                shutil.copyfile(self.image, self.image_file)
+
+            # Start the container instance.
+            self._runner.run(['singularity', 'instance.start',
+                self.image_file, self.name])
+
+        # Update status
+        self.id = self.name
+        self.status = 'running'
+        return {
+            'id': self.id,
+            'status': self.status,
+            'image_file': self.image_file
+        }
+
+    def delete(self):
+        """
+        Deletes a container instance.
+        """
+        # Check to see if the container is already stopped.
+        if not self.get_instance_info():
+            return
+
+        # Stop the container.
+        self._runner.run(['singularity', 'instance.stop', self.name])
+
+        # Update status
+        self.id = None
+        self.status = None
+
+        # Remove temporary image file
+        os.remove(self.image_file)
+
+    def start(self):
+        """
+        Start a stopped container.
+        """
+        lgr.info("Singularity does not provide a start feature.")
+
+    def stop(self):
+        """
+        Stop a running container.
+        """
+        lgr.info("Singularity does not provide a stop feature.")
+
+    @borrowdoc(Resource)
+    def get_session(self, pty=False, shared=None):
+        if not self.get_instance_info():
+            self.connect()
+
+        if pty and shared is not None and not shared:
```

Comment:
I'm confused about the values that `shared` can take.  `Resource.get_session` makes it seem like it can be None or a string.  Based on that, I'd guess that the intent of `shared is not None and not shared` is to check that it's not an empty string.  But grepping the code base, I see that `shared=False` is passed, so perhaps the meaning of False should be documented and then this could be just `shared is
False`?

Reply:
I believe the shared=False in resource.session.get_local_session() is an error. The parameter is defined as thus in base.Resource.get_session:

shared : string, optional
            Shared session identifier (the default is None)

Reply:
So are you guarding against an empty string?  If not, the `and not shared` can be dropped.

Reply:
Right, an empty string would not be a valid identifier. So I think that is what the *and not shared* is there to fail.